### ISSUE TITLE
Remote Input not working

### ISSUE THREAD

**Issue by theVivern** [andyholmes/gnome-shell-extension-gsconnect#329]
The remote input plugin does not work on my desktop PC. I can neither move the cursor nor produce mouse clicks.

The other plugins work fine, so it's definitely not the connection. In particular I can see the battery levels, switch between songs and view notifications no problem. I have the same phone connected to two laptops with similar configurations and remote input works there.

If there are any log files I can provide let me know.

---

Desktop & laptops:

- OS: Archlinux
- GNOME: 3.30.1 (running on X)
- GSConnect: v15

Phone:

- KDE Connect 1.10 (installed from F-Droid)
- Android 7.1.2

**Comment by andyholmes** [comment]
Hi, you've caught me just before I'm going to sleep, but I can get you started (thanks for the info, it's helpful).

There is an unfinished [Debugging](https://github.com/andyholmes/gnome-shell-extension-gsconnect/wiki/Debugging) page in the Wiki, if you follow that to get the debug log open you should be able to see traffic between the devices. There's going to be a lot of it (don't forget to turn it off; private info, filling up log files, etc), but what you're looking for is packets like these in the "GJS" tab:

```js
DEBUG: [core.js:receive/<:404]: Nexus 4: {
  "id": 1542112159112,
  "type": "kdeconnect.mousepad.request",
  "body": {
    "dx": 2,
    "dy": -7
  }
}
DEBUG: [core.js:receive/<:404]: Nexus 4: {
  "id": 1542112165066,
  "type": "kdeconnect.mousepad.request",
  "body": {
    "key": "a"
  }
}
```
That's a mouse movement and a key stroke. If it seems like you're getting those and there are no errors, then it's probably a problem with AT-SPI, which is usually the case. AT-SPI is usually for accessibility (screen readers, mouse keys, etc), but we use it simulate local events.

In this case you'll probably want to start looking in more general log files, using GNOME Logs is pretty easy. Probably there will be some errors from AT-SPI, I get them even when it's working, but try Googling any that seem suspicious and maybe compare them to your working computer. Let me know what you find and we'll go from there.

**Comment by theVivern** [comment]
It's exactly as you said. The log is full of packages like this one:

```JSON
DEBUG: [core.js:receive/<:404]: Satellite: {
  "id": 1542126667218,
  "type": "kdeconnect.mousepad.request",
  "body": {
    "dx": -3,
    "dy": 0
  }
}
```
I'm not on my Desktop right now, but I'll report back once if I find the problem. Chances are I'm only missing the AT-SPI package you mentioned.

---

On a side note: Clicking the "terminal" button in the debugging UI doesn't work for me. Turns out the terminal command is hardcoded as `gnome-terminal`. The button fails quietly for people that use a different terminal (like me). I'll add a short note on the Debugging wiki page if that's fine with you.

**Comment by andyholmes** [comment]
> I'll add a short note on the Debugging wiki page if that's fine with you.

Sure, the more important part is the switch, in that case. The button is just a convenience for opening `journalctl` for `gjs` and `gnome-shell`.

**Comment by ferdnyc** [comment]
Isn't "terminal application" typically one of the desktop-environment configurables? I wonder if there's any way we can get Gnome Shell to _tell_ us what command we should be using to launch a terminal window?

**Comment by ferdnyc** [comment]
Hmm. The GSettings schema `org.gnome.desktop.default-applications.terminal` contains `exec` and `exec-arg`, but I'm worried by the fact that mine are set to:
```console
$ gsettings list-recursively org.gnome.desktop.default-applications.terminal
org.gnome.desktop.default-applications.terminal exec-arg '-x'
org.gnome.desktop.default-applications.terminal exec 'gnome-terminal'
```

And the `-x` argument to `gnome-terminal` is so out of date that it's both (a) undocumented in the `--help` and (b) reported as deprecated if you actually use it:
```console
$ gnome-terminal -x
# Option “-x” is deprecated and might be removed in a later version of gnome-terminal.
# Use “-- ” to terminate the options and put the command line to execute after it.
# Failed to parse arguments: Unknown option -x
```
Aha, yeah:
```console
$ gsettings describe org.gnome.desktop.default-applications.terminal exec
Terminal program to use when starting applications that require one.

DEPRECATED: This key is deprecated and ignored. The default terminal is handled in GIO.
```
...OK, "handled in GIO". That's as helpful as ever, thanks GNOME. \*google,google* _Oh, **COME ON!**_

From [reddit]:(https://www.reddit.com/r/gnome/comments/2z1ai5/change_default_terminal_application/cpeu7sd/)
> nautilus-open-terminal uses gio to launch the terminal since 0.20 [0] and **gio currently has a list of terminals that it looks for in a hardcoded order [1][2].** [...]
>
> [0] https://git.gnome.org/browse/nautilus-open-terminal/commit/?id=ce56165ed58b225804f50a01dc101097b7892efd
> [1] https://git.gnome.org/browse/glib/tree/gio/gdesktopappinfo.c#n2473
> [2] https://bugzilla.gnome.org/show_bug.cgi?id=627943

Still, I assume that GIO method must be sufficient for the terminal @theVivern is using, so if it works then it could probably work for us. At ref #1 in that quote, `nautilus-open-terminal` uses `g_app_info_create_from_commandline(command_line, NULL, G_APP_INFO_CREATE_NEEDS_TERMINAL, ...)` to launch `/bin/sh` in a terminal window. 

Yeah, this works, in my testing.  I cribbed from `nautilus-open-terminal`. (Note that I changed the `spawn_command_line_async` command line to "gnome-term**b**inal", to ensure it would fail)
```js
        try {
          // Launch a terminal with tabs for GJS and GNOME Shell
          GLib.spawn_command_line_async(
            'gnome-termbinal ' +
            '--tab --title "GJS" --command "journalctl -f -o cat /usr/bin/gjs" ' +
            '--tab --title "GNOME Shell" --command "journalctl -f -o cat /usr/bin/gnome-shell"'
          );
        } catch(e) {
          logError(e);
          // OK, let's try the hard way around
          let disp = new Gdk.Display;
          let ctx = disp.get_app_launch_context();
          let app = Gio.AppInfo.create_from_commandline('journalctl -f -o cat /usr/bin/gjs', 'GJS', Gio.AppInfoCreateFlags.NEEDS_TERMINAL);
          app.launch([], ctx);
        }
```
(That still launches `gnome-terminal` on my system, since it's installed and available, but it does it through Gio in the I-guess-"proper" manner so that it should use _other_ terminal apps if `gnome-terminal` isn't installed.)

The two-tabs view will have to go if we can't launch `gnome-terminal` ourselves, obviously, since there's no generic way to indicate that. I'll just have it open two separate windows, I guess. I'll put together a PR that uses the above method to do fallback.

**Comment by ferdnyc** [comment]
@theVivern: Though merging the debugging fixes auto-closed this, whenever you next update GSConnect to a version > `v16`, could you check that the Debug Terminal button is now able to launch terminal-window log views for you? Thanks.

(I'm not sure exactly when `v17` will be releasing, and once it does it'll take some time to get reviewed and show up on Extensions.gnome.org. So it'll be a while before the next update. I'll try to remember to ping you once it's available.)

**Comment by ferdnyc** [comment]
Oops, and I just noticed this issue was about more than debug logs. Reopening.

**Comment by theVivern** [comment]
Sorry for the long delay.
Both systems have the same AT-SPI packages installed, namely `at-spi2-atk` and `extra/at-spi2-core`, so that can't be the problem. As I didn't perform any AT related configuration and both systems were reinstalled in the last couple months I doubt that this is the problem.

I found this error in `gnome-logs`:

> Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-M28BTZ/socket: No such file or directory

However, this error shows up on both systems as well. Is there a hardcoded socket that happens to work on one system and not the other?

**Comment by theVivern** [comment]
Sorry for the long delay.
Both systems have the same AT-SPI packages installed, namely `at-spi2-atk` and `extra/at-spi2-core`, so that can't be the problem. As I didn't perform any AT related configuration and both systems were reinstalled in the last couple months I doubt that this is the problem.

I found this error in `gnome-logs`:

> Unable to open bus connection: Failed to connect to socket /run/user/1000/at-spi2-M28BTZ/socket: No such file or directory

However, this error shows up on both systems as well. Is there a hardcoded socket that happens to work on one system and not the other?

**Comment by theVivern** [comment]
> @theVivern: Though merging the debugging fixes auto-closed this, whenever you next update GSConnect to a version > `v16`, could you check that the Debug Terminal button is now able to launch terminal-window log views for you? Thanks.

I'm on gsconnect v17 now and the behaviour hasn't changed for me. The button works as expected when `gnome-terminal` is installed. Without it there's this error message in `gnome-logs`:

> JS ERROR: GLib.SpawnError: Failed to execute child process “gnome-terminal” (No such file or directory)
_logAction@/home/jakob/.local/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/ui/debug.js:351:13
@/home/jakob/.local/share/gnome-shell/extensions/gsconnect@andyholmes.github.io/service/daemon.js:839:2

 My terminal emulator is `tilix`. Very like it's just not on gio's list of terminals that was mentioned above.

**Comment by andyholmes** [comment]
> Sorry for the long delay.
> Both systems have the same AT-SPI packages installed, namely at-spi2-atk and extra/at-spi2-core, so that can't be the problem.

No problem. I'm not sure why it would work on one but not the other.

> I found this error in `gnome-logs`...

I get those too, but I'm not sure if/how critical they are. AT-SPI seems to open new sockets frequently. If you run `ls /run/user/1000/` you'll probably see a lot of old sockets left around.

I believe this socket is just a secured DBus socket. AT-SPI is really just a DBus service, there's just a convenient API in Gnome for using it without making raw DBus calls. You can try running these commands by starting `gjs-console` in a terminal. It might give some better clue.

```js
let Atspi = imports.gi.Atspi;
Atspi.init();
Atspi.generate_keyboard_event(0, 'a', 4)
```

Output should look like this:

```sh
$ gjs-console
gjs> let Atspi = imports.gi.Atspi;
gjs> Atspi.init();
0
gjs> Atspi.generate_keyboard_event(0, 'a', 4);
atrue
gjs>
```

> My terminal emulator is tilix. Very like it's just not on gio's list of terminals that was mentioned above.

Looks like there's some [instructions for that here](https://github.com/gnunn1/tilix/issues/863).

**Comment by theVivern** [comment]
After pondering the differences between my laptop and desktop some more I figured out the problem. Turns out, the plugin works when I start GNOME via GDM. I used to use `startx` (called by a script) on my deskop, but use GDM on my laptop because it is required for the lock screen to work.

---
> Output should look like this:

For the record, this is my output:

```
gjs> let Atspi = imports.gi.Atspi;
gjs> Atspi.init();
0
gjs> Atspi.generate_keyboard_event(0, 'a', 4)
true
gjs>
```

There's also this error in the logs:
> Keystring synthesis failure, string=a
